File: lib/python/src/aria/ops/adapter_instance.py
#  Copyright 2022 VMware, Inc.
#  SPDX-License-Identifier: Apache-2.0
from __future__ import annotations

import sys
from typing import Dict
from typing import Optional

from aria.ops.certificate_info import CertificateInfo
from aria.ops.object import Identifier
from aria.ops.object import Key
from aria.ops.object import Object
from aria.ops.pipe_utils import read_from_pipe
from aria.ops.suite_api_client import SuiteApiClient
from aria.ops.suite_api_client import SuiteApiConnectionParameters


class AdapterInstance(Object):  # type: ignore
    def __init__(self, json: dict) -> None:
        adapter_key = json.get("adapter_key", {})
        super().__init__(
            Key(
                adapter_kind=adapter_key.get("adapter_kind"),
                object_kind=adapter_key.get("object_kind"),
                name=adapter_key.get("name"),
                identifiers=[
                    Identifier(
                        identifier.get("key"),
                        identifier.get("value"),
                        identifier.get("is_part_of_uniqueness"),
                    )
                    for identifier in adapter_key.get("identifiers", [])
                ],
            )
        )

        credential_config = json.get("credential_config")
        if type(credential_config) is dict:
            self.credential_type = credential_config.get("credential_key", None)
            self.credentials = {
                credential.get("key"): credential.get("value")
                for credential in credential_config.get("credential_fields", [])
            }
        else:
            self.credential_type = None
            self.credentials = {}

        cluster_connection_info = json.get("cluster_connection_info")
        if type(cluster_connection_info) is dict:
            self.suite_api_client = SuiteApiClient(
                SuiteApiConnectionParameters(
                    username=cluster_connection_info.get("user_name"),
                    password=cluster_connection_info.get("password"),
                    host=cluster_connection_info.get("host_name"),
                )
            )
        else:
            self.suite_api_client = None

        certificate_config = json.get("certificate_config")
        self.certificates: list[CertificateInfo] = []
        if type(certificate_config) is dict:
            self.certificates = [
                CertificateInfo(cert)
                for cert in certificate_config.get("certificates", [])
            ]

        self.collection_number: Optional[int] = json.get("collection_number", None)
        self.collection_window: Optional[Dict] = json.get("collection_window", None)

    def get_suite_api_client(self) -> Optional[SuiteApiClient]:
        """
        Gets an authenticated Suite API Client. Note that Suite API calls can only be
        made from the 'collect' endpoint. Returns 'None' when called from 'test' and
        'get_endpoints'.

        Returns:
            Authenticated SuiteApiClient, or None
        """
        return self.suite_api_client

    def get_certificates(self) -> list[CertificateInfo]:
        """
        Gets a list of all certificates that have been validated by a CA or manually
        accepted by a user.

        Returns:
            A list of certificates for verifying SSL connections.
        """
        return self.certificates

    def get_collection_number(self) -> Optional[int]:
        """
        Gets the current collection number, starting from 0.
        Returns 'None' when called from 'test' and 'get_endpoints'.

        Returns:
            The current collection number
        """
        return self.collection_number

    def get_collection_window(self) -> Optional[Dict]:
        """Gets the window for the current collection.

         In some cases, it is useful to have a start and end time for collection. The
         start_time and end_time will be modified each collection such that each
         collection's end_time will be the next collection's start_time. Thus, the
         window can be treated as either the interval `(start_time, end_time]` or
         `[start_time, end_time)`, so long as each collection uses the same convention,
         there will be no overlaps or missing times. (Note that restarting the adapter
         instance will reset the time window. In general, this will cause overlap
         and/or missing times when restarts occur for any reason).

         Times are provided in milliseconds since the Epoch.

        Returns:
            A dictionary with two keys: 'start_time' and 'end_time'. On the first
            collection, 'start_time' will be set to '0'. On calls to 'test' and
            'get_endpoints', this will be 'None'.
        """
        return self.collection_window

    def get_credential_type(self) -> Optional[str]:
        """Get the type (key) of credential. This is useful if an adapter supports multiple types of credentials.

        Returns:
            the type of the credential used by this adapter instance, or None if the adapter instance does not have a credential.
        """
        return self.credential_type  # type: ignore[no-any-return]

    def get_credential_value(self, credential_key: str) -> Optional[str]:
        """Retrieve the value of a given credential

        Args:
            credential_key (str): Key of the credential field

        Returns:
            value associated with the credential field, or None if a credential field with the given key does not exist.
        """
        return self.credentials.get(credential_key)

    @classmethod
    def from_input(cls, infile: str = sys.argv[-2]) -> AdapterInstance:
        # The server always invokes methods with the input file as the second to last argument
        return cls(read_from_pipe(infile))

File: lib/java/AdapterLibrary/src/main/kotlin/com/vmware/aria/operations/AdapterInstance.kt
/*
 * Copyright 2023 VMware, Inc.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.vmware.aria.operations

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

@Serializable
data class CredentialField(
    val key: String,
    val value: String,
    @SerialName("is_password") val isPassword: Boolean,
)

@Serializable
class Credential(
    /**
     * Get the type (key) of credential. This is useful if an adapter supports multiple
     * types of credentials.
     *
     * @return The type of the credential used by this adapter instance, or null if the
     * adapter instance does not have a credential.
     */
    @SerialName("credential_key") val type: String?,
    @SerialName("credential_fields") private val fields: List<CredentialField>,
) : Map<String, String> by fields.associate({ field -> Pair(field.key, field.value) })

/**
 * Class that represents a list of validated SSL certificates that have been verified
 * automatically by a CA or manually by the user.
 */
@Serializable
class Certificates(private val certificates: List<String>) :
    List<String> by certificates

/**
 * Class that describes a time window. In some cases, it is useful to have a start and end time
 * for collection. The CollectionWindow's [startTime] and [endTime] will be modified each collection
 * such that each collection's [endTime] will be the next collection's [startTime]. Thus, the window
 * can be treated as either the interval `(startTime, endTime]` or `[startTime, endTime)`, so long
 * as each collection uses the same convention, there will be no overlaps or missing times. (Note
 * that restarting the adapter instance will reset the time window. In general, this will cause
 * overlap and/or missing times when restarts occur for any reason)
 */
@Serializable
data class CollectionWindow(
    /**
     * The start of the window. On the first collection, this will be set to `0`.
     */
    @SerialName("start_time") val startTime: Double,
    /**
     * The end of the window
     */
    @SerialName("end_time") val endTime: Double,
)

class AdapterInstance(json: JsonObject) : Object(getKey(json)) {
    /**
     * Gets information about the credential (if it exists) for this Adapter Instance
     */
    val credential: Credential

    /**
     * Gets an authenticated [Suite API Client][SuiteApiClient].
     */
    val suiteApiClient: SuiteApiClient

    /**
     * Gets the list of SSL Certificates that have been verified automatically by a CA
     * or manually by the user.
     */
    val certificates: Certificates

    /**
     * Gets the current collection number, starting from 0.
     */
    val collectionNumber: Int

    /**
     * Gets the current collection window.
     */
    val collectionWindow: CollectionWindow

    init {
        credential = json["credential_config"]?.let {
            Json.decodeFromJsonElement(it)
        } ?: Credential(null, emptyList())

        suiteApiClient = SuiteApiClient(
            when (val suiteApiConnectionInfo = json["cluster_connection_info"]) {
                is JsonNull, null -> SuiteApiConnectionInfo("", "", "")
                else -> Json.decodeFromJsonElement<SuiteApiConnectionInfo>(
                    suiteApiConnectionInfo
                )
            }
        )

        certificates = json["certificate_config"]?.let {
            Json.decodeFromJsonElement(it)
        } ?: Certificates(emptyList())

        collectionNumber = json["collection_number"]?.jsonPrimitive?.intOrNull ?: 0
        collectionWindow = json["collection_window"]?.let {
            Json.decodeFromJsonElement(it)
        } ?: CollectionWindow(0.0, System.currentTimeMillis().toDouble())
    }

    /**
     * @return The type (key) of the credential, or `null` if this Adapter Instance does
     * not have a credential.
     */
    fun getCredentialType(): String? = credential.type

    /**
     * Retrieve the value of a given credential
     * @param key Key of the credential field
     * @return value associated with the credential field, or null if a credential field
     * with the given key does not exist.
     */
    fun getCredentialValue(key: String): String? = credential[key]

    companion object {
        /**
         * Create an [AdapterInstance] from the input pipe.
         */
        @JvmStatic
        @JvmOverloads
        fun fromInput(inputFile: String = Pipes.input): AdapterInstance {
            val json = readFromPipe(inputFile)
                ?: throw IllegalArgumentException("Cannot read Adapter Instance from input")
            return AdapterInstance(json.jsonObject)
        }


        private fun getKey(json: JsonObject): Key {
            return json["adapter_key"]?.jsonObject?.let { adapterKey ->
                Key(
                    adapterType = adapterKey["adapter_kind"]?.jsonPrimitive?.content
                        ?: "",
                    objectType = adapterKey["object_kind"]?.jsonPrimitive?.content
                        ?: "",
                    name = adapterKey["name"]?.jsonPrimitive?.content ?: "",
                    identifiers = adapterKey["identifiers"]?.jsonArray?.map { identifier ->
                        Identifier(
                            identifier.jsonObject["key"]?.jsonPrimitive?.content ?: "",
                            identifier.jsonObject["value"]?.jsonPrimitive?.content
                                ?: "",
                            identifier.jsonObject["is_part_of_uniqueness"]?.jsonPrimitive?.boolean
                                ?: true
                        )
                    } ?: listOf()
                )
            }
                ?: throw IllegalArgumentException("Cannot read Adapter Instance Key from input")
        }
    }
}
